src/v/cloud_storage/remote.cc

```diff
@@ -137,473 +139,607 @@ static error_outcome categorize_error(
-            retry_permit = fib.retry();
-            break;
-        case error_outcome::fail:
-            result = download_result::failed;
-            vlog(
-              ctxlog.warn,
-              "Failed downloading manifest from {} {}, manifest at {}",
-              bucket,
-              *result,
-              path);
+        static_cast<Derived*>(this)->on_timedout(bucket, result, key);
+        co_return *result;
+    }
+
+    retry_chain_logger& ctxlog() { return _ctxlog; }
+
+private:
+    const ss::gate& _gate;
+    ss::gate::holder _guard;
+    retry_chain_node _fib;
+    retry_chain_logger _ctxlog;
+    retry_permit _retry_permit;
+    s3::client_pool& _pool;
+    ss::abort_source& _as;
+};
+
+struct retriable_manifest_download
+  : retriable<retriable_manifest_download, download_result> {
+    retriable_manifest_download(
+      ss::gate& g,
+      retry_chain_node& parent,
+      s3::client_pool& pool,
+      ss::abort_source& as,
+      remote_probe& probe,
+      base_manifest& m)
+      : retriable<retriable_manifest_download, download_result>(
+        g, parent, pool, as)
+      , _probe(probe)
+      , _manifest(m) {}
+
+    ss::future<download_result> do_ivoke(
+      s3::client_pool::http_client_ptr client,
+      const s3::bucket_name& bucket,
+      const s3::object_key& key,
+      ss::lowres_clock::duration timeout) {
+        auto resp = co_await client->get_object(bucket, key, timeout);
+        vlog(ctxlog().debug, "Receive OK response from {}", key);
+        co_await _manifest.update(resp->as_input_stream());
+        switch (_manifest.get_manifest_type()) {
+        case manifest_type::partition:
+            _probe.partition_manifest_download();
             break;
-        case error_outcome::notfound:
-            result = download_result::notfound;
-            vlog(
-              ctxlog.debug,
-              "Manifest from {} {}, manifest at {} not found",
-              bucket,
-              *result,
-              path);
+        case manifest_type::topic:
+            _probe.topic_manifest_download();
             break;
         }
+        co_return download_result::success;
     }
-    _probe.failed_manifest_download();
-    if (!result) {
+
+    void on_backoff(
+      const s3::bucket_name& bucket,
+      std::optional<result_t>&,
+      const s3::object_key&,
+      const retry_permit& permit) {
+        _probe.manifest_download_backoff();
         vlog(
-          ctxlog.warn,
-          "Downloading manifest from {}, backoff quota exceded, manifest at {} "
-          "not available",
+          ctxlog().debug,
+          "Downloading manifest from {}, {} backoff required",
           bucket,
-          path);
-        result = download_result::timedout;
+          std::chrono::milliseconds(permit.delay));
     }
-    co_return *result;
-}
 
-ss::future<upload_result> remote::upload_manifest(
-  const s3::bucket_name& bucket,
-  const base_manifest& manifest,
```

Comment:
I'm wondering if there is a way to either (1) split up this patch or (2) more likely to work, move the new/factored-out code some where else in the file (e.g. to the bottom) which doesn't result in a diff where so many +/- lines are interleaved. For example, it looks like the diff algorithm is interleaving the removal of the `remote::*` bits with the addition of `class retriable`.

I'm definitely sympathetic to the complexity in this patch being inherent with no good way to make it better, but thought it was worth asking.